File: easy/ts/flood_fill.ts
function floodFill(image: number[][], sr: number, sc: number, color: number): number[][] {
  const origin: number = image[sr][sc]

  function fill(image: number[][], sr: number, sc: number): number[][] {
    if (sr < 0 || image.length <= sr || sc < 0 || image[0].length <= sc) return image
    if (image[sr][sc] !== origin) return image
    if (image[sr][sc] === color) return image

    image[sr][sc] = color

    fill(image, sr - 1, sc)
    fill(image, sr, sc - 1)
    fill(image, sr + 1, sc)
    fill(image, sr, sc + 1)

    return image
  }

  fill(image, sr, sc)

  return image
};

File: easy/ruby/flood_fill.rb
# @param {Integer[][]} image
# @param {Integer} sr
# @param {Integer} sc
# @param {Integer} color
# @return {Integer[][]}
def flood_fill(image, sr, sc, color)
  origin = image[sr][sc]
  fill = lambda do |r, c|
    return if r.negative? || image.length <= r ||
              c.negative? || image[0].length <= c

    return if image[r][c] != origin || image[r][c] == color

    image[r][c] = color

    fill.call(r, c + 1)
    fill.call(r, c - 1)
    fill.call(r + 1, c)
    fill.call(r - 1, c)
  end

  fill.call(sr, sc)
  image
end
